File: app/src/main/java/dev/favourdevlabs/cleanthes/utils/CleanthesApplication.java
package dev.favourdevlabs.cleanthes.utils;

import android.app.Application;
import android.app.Activity;
import android.os.Bundle;

import dev.favourdevlabs.cleanthes.ui.auth.SessionManager;
import dev.favourdevlabs.cleanthes.utils.ClipboardHelper;

public class CleanthesApplication extends Application {

    private int startedActivityCount = 0;

    @Override
    public void onCreate() {
        super.onCreate();
        registerActivityLifecycleCallbacks(new AppLifecycleTracker());
    }

    private class AppLifecycleTracker implements ActivityLifecycleCallbacks {

        @Override
        public void onActivityStarted(Activity activity) {
            startedActivityCount++;
        }

        @Override
        public void onActivityStopped(Activity activity) {
            startedActivityCount--;

            if (startedActivityCount == 0) {
                onAppBackgrounded();
            }
        }

        @Override
        public void onActivityCreated(Activity a, Bundle b) {
        }

        @Override
        public void onActivityResumed(Activity a) {
        }

        @Override
        public void onActivityPaused(Activity a) {
        }

        @Override
        public void onActivitySaveInstanceState(Activity a, Bundle b) {
        }

        @Override
        public void onActivityDestroyed(Activity a) {
        }
    }

    private void onAppBackgrounded() {
        SessionManager.clearSession();
        ClipboardHelper.clearClipboard(getApplicationContext());
    }
}

File: app/src/main/java/dev/favourdevlabs/cleanthes/utils/CleanthesApplication.kt
package dev.favourdevlabs.cleanthes.utils

import android.app.Activity
import android.app.Application
import android.os.Bundle
import dev.favourdevlabs.cleanthes.ui.auth.SessionManager

class CleanthesApplication : Application() {

    private var startedActivityCount = 0

    override fun onCreate() {
        super.onCreate()
        registerActivityLifecycleCallbacks(AppLifecycleTracker())
    }

    private inner class AppLifecycleTracker : ActivityLifecycleCallbacks {

        override fun onActivityStarted(activity: Activity) {
            startedActivityCount++
        }

        override fun onActivityStopped(activity: Activity) {
            startedActivityCount--
            if (startedActivityCount == 0) onAppBackgrounded()
        }

        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
        override fun onActivityResumed(activity: Activity)                              = Unit
        override fun onActivityPaused(activity: Activity)                               = Unit
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle)  = Unit
        override fun onActivityDestroyed(activity: Activity)                            = Unit
    }

    private fun onAppBackgrounded() {
        SessionManager.clearSession()
        ClipboardHelper.clearClipboard(applicationContext)
    }
}
